### PR TITLE
oop-exercise

### PR DESCRIPTION
**Pull request by ik-web**
### oop-exercise

[Demo](https://ik-web.github.io/a-tiny-JS-world/)
[Code base](https://github.com/ik-web/a-tiny-JS-world)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

**Comment by OleksiyRudenko** [comment]
Please review updated submission instructions.
Not a problem here but new instructions offer a better workflow

File: submissions/ik-web/oop-exercise/index.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+*/
+
+// ======== OBJECTS DEFINITIONS ========
+class Inhabitant {
+    constructor(species, name, gender, legs, hands, saying) {
+       this.species = species;
+       this.name = name;
+       this.gender = gender;
+       this.legs = legs;
+       this.hands = hands;
```
Comment (OleksiyRudenko) [inline submissions/ik-web/oop-exercise/index.js:16]: Not all inhabitants have hands by natural design. The beauty of OOP is that every base class defines and handles only the props that are shared by its descendants, and every child class introduces and handles props that are unique to it and its descendants.

File: submissions/ik-web/oop-exercise/index.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ const man = new Human('Benjamin', 'male', 'Hi to all!');
+ const woman = new Human('Sarah', 'female', 'Hello everyone!');
+ const catWoman = new CatWoman('anonymos', cat.getSaying());
+ const tinyJsWorldPopulation = [dog, cat, man, woman, catWoman];
+ 
+ function toPrint(arr) {
+    arr.forEach( function(obj) {
+       print(`A ${obj.species} (<i>${obj.gender}<i>) named <strong>${obj.name}</strong> has ${obj.legs} legs and ${obj.hands} hands, likes to say <u>${obj.saying}</u>`);
+    });
+ }
```
Comment (OleksiyRudenko) [inline submissions/ik-web/oop-exercise/index.js:60]: Let classes handle their own representation. This is one of the points of OOP/SOLID principles.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@ik-web pretty nice job done.
However there is still some room for improvement

File: submissions/ik-web/oop-exercise/index.js (hunk 2 above)
Comment (ik-web) [reply submissions/ik-web/oop-exercise/index.js]: While working on this edit, I came across the fact that my constructor for a catwoman and a human is very similar and I find it difficult to decide ... leave how I did it or try to inherit a catwoman from a human, I can't accept the correct solution ((

File: submissions/ik-web/oop-exercise/index.js (hunk 1 above)
Comment (ik-web) [reply submissions/ik-web/oop-exercise/index.js]: I fixed this moment

File: submissions/ik-web/oop-exercise/index.js (hunk 2 above)
Comment (ik-web) [reply submissions/ik-web/oop-exercise/index.js]: Fixed

**Comment by zonzujiro** [comment]
@OleksiyRudenko can we merge this? :)

File: submissions/ik-web/oop-exercise/index.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      super(name, 'female', 2, saying);
+      this.species = 'mystery of nature';
+   }
+}
+
+const dog = new Dog('Bobik', 'male');
+const cat = new Cat('Barsik', 'male');
+const man = new Human('Benjamin', 'male', 2, 'Hi to all!');
+const woman = new Human('Sarah', 'female', 2, 'Hello everyone!');
+const catWoman = new CatWoman('anonymos', cat.getSaying());
```
Comment (OleksiyRudenko) [inline submissions/ik-web/oop-exercise/index.js:62]: Here a cat-woman borrows saying from a particular cat. What if we've got more than 1 cat? What if we've got no cats at all?
If it was supposed to borrow saying from any cat default saying then [class static properties and methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/static) are to the rescue.

**Review by OleksiyRudenko — APPROVED** [review]
@ik-web approved with a reservation.
Take a minute to reflect on the comment below.

File: submissions/ik-web/oop-exercise/index.js (hunk 3 above)
Comment (ik-web) [reply submissions/ik-web/oop-exercise/index.js:62]: Thank you for the help and advice

**Pull request merged**